### PR TITLE
nixos/prometheus-lnd-exporter: init

### PR DESCRIPTION
**Pull request by mmilata**
###### Motivation for this change

fort-nix/nix-bitcoin#186 + would like to take advantage of the nice prometheus exporters infrastructure that nixpkgs has

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by mdlayher — APPROVED** [review]
LGTM with a test nit.

File: nixos/tests/prometheus-exporters.nix [hunk 1]
```diff
@@ -202,6 +202,46 @@ let
...
+          wantedBy = [ "multi-user.target" ];
+          after = [ "network.target" ];
+        };
+      };
+      exporterTest = ''
+        wait_for_unit("lnd.service")
+        wait_for_open_port(10009)
+        wait_for_unit("prometheus-lnd-exporter.service")
+        wait_for_open_port(9092)
+        succeed("curl -sSf localhost:9092/metrics | grep -q '^promhttp_metric_handler'")
```
[the remark this answers is not in the document]
Comment (mmilata) [reply nixos/tests/prometheus-exporters.nix:241]: It seems that in order for an actual metric to be exported the lnd daemon needs to be synced with bitcoind and have wallet generated. Although it should be possible to test this without internet access by running both daemons in "regtest" mode, the setup is quite non-trivial:(

File: nixos/modules/services/monitoring/prometheus/exporters/lnd.nix [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      ${pkgs.prometheus-lnd-exporter}/bin/lndmon \
+        --prometheus.listenaddr=${cfg.listenAddress}:${toString cfg.port} \
+        --prometheus.logdir=/var/log/prometheus-lnd-exporter \
+        --lnd.host=${cfg.lndHost} \
+        --lnd.tlspath=${cfg.lndTlsPath} \
+        --lnd.macaroondir=${cfg.lndMacaroonDir} \
+        ${concatStringsSep " \\\n  " cfg.extraFlags}
+    '';
+    serviceConfig.LogsDirectory = "prometheus-lnd-exporter";
+  };
```
Comment (WilliButz) [inline nixos/modules/services/monitoring/prometheus/exporters/lnd.nix:44]:
```suggestion
  serviceOpts.serviceConfig = {
    ExecStart = ''
      ${pkgs.prometheus-lnd-exporter}/bin/lndmon \
        --prometheus.listenaddr=${cfg.listenAddress}:${toString cfg.port} \
        --prometheus.logdir=/var/log/prometheus-lnd-exporter \
        --lnd.host=${cfg.lndHost} \
        --lnd.tlspath=${cfg.lndTlsPath} \
        --lnd.macaroondir=${cfg.lndMacaroonDir} \
        ${concatStringsSep " \\\n  " cfg.extraFlags}
    '';
    LogsDirectory = "prometheus-lnd-exporter";
    ReadOnlyPaths = "${cfg.lndTlsPath} ${lndMacaroonDir}";
  };
```

File: nixos/tests/prometheus-exporters.nix [hunk 3]
```diff
@@ -202,6 +202,46 @@ let
       '';
     };
 
+    lnd = {
+      exporterConfig = {
+        enable = true;
+        user = "root";
```
Comment (WilliButz) [inline nixos/tests/prometheus-exporters.nix:208]:
```suggestion
```

**Review by WilliButz — CHANGES_REQUESTED** [review]
I think running the exporter as root in the test  or as a normal user anywhere else can be avoided by giving it access to the needed files specifying `ReadOnlyPaths` (or if necessary `ReadWritePaths`) in the service configuration.
(See `systemd.exec(5)`)

Otherwise this looks good to me :+1:

File: nixos/tests/prometheus-exporters.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (WilliButz) [reply nixos/tests/prometheus-exporters.nix:241]: If it is too much overhead to actually bootstrap an environment for the exporter to work as intended, I think it is fine to only check if the module is compatible with the current exporter package. This should make it less likely that someone updates the exporter package without touching the module and thereby unknowingly causing incompatibilities, which has happened before :).

**Comment by mmilata** [comment]
Thanks @WilliButz, I've applied your suggestions and rebased on master to resolve conflicts.

**Comment by WilliButz** [comment]
@GrahamcOfBorg test prometheus-exporters.lnd

**Review by WilliButz — APPROVED** [review]
looks good to me :+1:

**Comment by mmilata** [comment]
Thanks @WilliButz and @mdlayher!

**Pull request merged**